iot_verifier/src/rewarder.rs

```diff
@@ -141,3 +274,108 @@ async fn save_rewarded_timestamp(
 ) -> db_store::Result<()> {
     meta::store(db, timestamp_key, value.timestamp()).await
 }
+
+pub async fn fetch_halt_rewards_override(db: impl sqlx::PgExecutor<'_>) -> db_store::Result<bool> {
+    Ok(sqlx::query_scalar::<_, bool>(
+        r#"
+    SELECT EXISTS(SELECT value from meta where key = $1 and value = 'true')
+    "#,
+    )
+    .bind(HALT_REWARDS_DB_KEY)
+    .fetch_one(db)
+    .await?)
+}
+
+async fn clear_halt_rewards_override(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> db_store::Result<()> {
+    meta::store(txn, HALT_REWARDS_DB_KEY, "false").await
+}
+
+async fn save_history(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    total_beacon_count: u64,
+    total_witness_count: u64,
+    total_packet_count: u64,
+    epoch_end: DateTime<Utc>,
+) -> anyhow::Result<()> {
+    history_db::insert(
+        txn,
+        RewardType::Beacon,
+        total_beacon_count as i64,
+        epoch_end,
+    )
+    .await?;
+    history_db::insert(
+        txn,
+        RewardType::Witness,
+        total_witness_count as i64,
+        epoch_end,
+    )
+    .await?;
+    history_db::insert(
+        txn,
+        RewardType::Packet,
+        total_packet_count as i64,
+        epoch_end,
+    )
+    .await?;
+    Ok(())
+}
+
+mod history_db {
+    use super::*;
+
+    pub async fn insert(
+        tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+        reward_type: RewardType,
+        count: i64,
+        epoch_ts: DateTime<Utc>,
+    ) -> anyhow::Result<()> {
+        sqlx::query(
+            r#"
+        INSERT INTO rewards_history(reward_type, count, epoch_ts) VALUES($1, $2, $3)
+        "#,
+        )
+        .bind(reward_type)
+        .bind(count)
+        .bind(epoch_ts)
+        .execute(tx)
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn get_avg(
+        db: impl sqlx::PgExecutor<'_>,
+        reward_type: RewardType,
+        default: u64,
+    ) -> anyhow::Result<u64> {
+        let result = sqlx::query_scalar::<_, Decimal>(
+            r#"
+        SELECT COALESCE(AVG(count), 0) FROM rewards_history where reward_type = $1
+        "#,
+        )
+        .bind(reward_type)
+        .fetch_one(db)
+        .await?;
+        let result = (result * AVG_SCALING_FACTOR).to_u64().unwrap_or(default);
+        if result < default {
+            Ok(default)
+        } else {
+            Ok(result)
+        }
+    }
```

Comment:
```suggestion
    pub async fn get_avgs(
        db: impl sqlx::PgExecutor<'_>,
        default_beacon: u64,
        default_witness: u64,
        default_packet: u64
    ) -> anyhow::Result<(u64, u64, u64)> {
        let row = sqlx::query(
            r#"
        SELECT COALESCE(AVG(count), 0) FROM rewards_history where reward_type = 'beacon' as beacon_avg,
            COALESCE(AVG(count), 0) FROM rewards_history where reward_type = 'witness' as witness_avg,
            COALESCE(AVG(count), 0) FROM rewards_history where reward_type = 'packet' as packet_avg,
        "#,
        )
        .fetch_one(db)
        .await?;
        let beacon_avg = row.try_get::<Decimal, &str>("beacon_avg").map(|ba| (ba * AVG_SCALING_FACTOR).to_u64()).unwrap_or(default_beacon);
        let witness_avg = row.try_get::<Decimal, &str>("witness_avg").map(|wa| (wa * AVG_SCALING_FACTOR).to_u64()).unwrap_or(default_witness);
        let packet_avg = row.try_get::<Decimal, &str>("packet_avg").map(|pa| (pa * AVG_SCALING_FACTOR).to_u64()).unwrap_or(default_packet);
        Ok(
            std::cmp::max(beacon_avg, default_beacon),
            std::cmp::max(witness_avg, default_witness),
            std::cmp::max(packet_avg, default_packet),
        )
    }
```
you should be able to get the averages out with a single query and simplify that calculation as well. may be able to do it _even_ more simply by querying against the rewards_history table with a single select and using group_by but not sure about that one

Reply:
i had something like that in an earlier version and reverted to individual queries but i think the code reads better as it is currently and its not like this flow needs to be optimised.  The queries are run once per day under normal conditions or once per hour for a fixed number of hours when rewards are suspended